sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala

```diff
@@ -69,6 +70,23 @@ private[hive] case class HiveSimpleUDF(
     udfType != null && udfType.deterministic() && !udfType.stateful()
   }
 
+  override def inputTypes: Seq[AbstractDataType] = {
+    val inTypes = children.map(_.dataType)
```

Comment:
I'm confused. The expected type should be defined by the function signature, but not the actual function inputs. What are we doing here?

As an example, `ScalaUDF.inputTypes` is derived from the function signature (captured by encoders).

Reply:
The `ScalaUDF` also implicit convert input type to expected type at `ImplicitTypeCasts`. 

Let's say we have a udf `Array[Double] => Double = { data => data.sum }` and we run `spark.sql("select udf(array(1.0, 1.1, 1.2))")`. Then `ImplicitTypeCasts` will cast `array<decimal>` to `array<double>`.

But Hive udf can't enjoy it, now we only use Hive `ObjectInspector` to convert data type at running time. It's the difference.

Reply:
Ah I misread the code. We get children data type only to skip decimal.